google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java

```diff
@@ -145,7 +148,35 @@ public static CloudStorageFileSystem forBucket(String bucket, CloudStorageConfig
         !bucket.startsWith(URI_SCHEME + ":"), "Bucket name must not have schema: %s", bucket);
     checkNotNull(config);
     return new CloudStorageFileSystem(
-        new CloudStorageFileSystemProvider(config.userProject()), bucket, config);
+        getCloudStorageFileSystemProvider(config, null), bucket, config);
+  }
+
+  /**
+   * Creates new file system provider instance for {@code CloudStorageFileSystemProvider}.
+   *
+   * @param config CloudStorageConfiguration to initialize {@code CloudStorageFileSystemProvider}
+   *     object
+   * @param storageOptions
+   * @return {@code CloudStorageFileSystemProvider} object
+   */
+  private static CloudStorageFileSystemProvider getCloudStorageFileSystemProvider(
+      CloudStorageConfiguration config, StorageOptions storageOptions) {
+    CloudStorageFileSystemProvider cloudStorageFileSystemProvider =
+        (storageOptions == null)
+            ? new CloudStorageFileSystemProvider(config.userProject())
+            : new CloudStorageFileSystemProvider(config.userProject(), storageOptions);
+    Set<CloudStorageFileSystemProvider> cloudStorageFileSystemProviderSet = new HashSet<>();
+    if (CONFIG_TO_PROVIDERS_MAP.get(config) != null) {
+      cloudStorageFileSystemProviderSet = CONFIG_TO_PROVIDERS_MAP.get(config);
+      for (CloudStorageFileSystemProvider provider : cloudStorageFileSystemProviderSet) {
+        if (provider.equals(cloudStorageFileSystemProvider)) {
+          cloudStorageFileSystemProvider = provider;
+        }
+      }
+    }
+    cloudStorageFileSystemProviderSet.add(cloudStorageFileSystemProvider);
+    CONFIG_TO_PROVIDERS_MAP.put(config, cloudStorageFileSystemProviderSet);
+    return cloudStorageFileSystemProvider;
```

Comment:
The logic you use is a bit complicated.
the more clear would be clear:

Set set = map.get();
if (set == null) {
   set = new HashSet();
   set.add(provider);
   map.put(config, set);
} else {
   for( existiningProvider: set) {
       if (existiningProvider.equals(newProvider)) return existingProvider;
    }
    // not found
   set.add(newProvider);
   return newProvider;
}

Reply:
Done